### PR TITLE
Support specifying mode when create new file by Alluxio FUSE

### LINKED ISSUE #9367: Support specifying mode when create new file by Alluxio FUSE
**Alluxio Version:**
Alluxio 1.8 or 2.0

**Describe the bug**
If create a new file with mode 0740, then get mode 0644.

**To Reproduce**
test program(C):
```
/* creat.c */
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <errno.h>

int main(int argc, const char **argv)
{
    //int ret = open("mode_0740", O_CREAT, S_IRWXU | S_IRGRP);
    int ret = creat("mode_0740", S_IRWXU | S_IRGRP);
    if (ret >= 0) {
        printf("open mode_0740 fd: %d\n", ret);
        return 0;
    } else {
        printf("open mode_0740 failed: %d\n", errno);
        return -errno;
    }
}
```
gcc creat.c -o /tmp/creat
cd /your/alluxio/fuse/directory
\# execute test program
/tmp/creat
ls -l mode_0740
-rw-r--r-- 1 root root 0 Jun 27 16:20 mode_0740
\# this mode_0740 file's mode is 0644

### PR DESCRIPTION
**Pull request by liuhongtong**
Fix #9367
Signed-off-by: liuhongtong <hongtongliu@126.com>

### CODE REVIEW

File: core/transport/src/main/java/alluxio/grpc/PMode.java [hunk 1]
```diff
@@ -675,5 +675,16 @@ public PMode parsePartialFrom(
     return DEFAULT_INSTANCE;
   }
 
+  static public Bits parseOwnerBits(long mode) {
```
Comment (apc999) [inline core/transport/src/main/java/alluxio/grpc/PMode.java]: This file `PMode.java` is auto generated by the protocol buffer compiler which we should not touch

File: integration/fuse/src/main/java/alluxio/fuse/AlluxioFuseFileSystem.java [hunk 2]
```diff
@@ -246,8 +248,13 @@ public int create(String path, @mode_t long mode, FuseFileInfo fi) {
...
       }
-
-      FileOutStream os = mFileSystem.createFile(uri);
+      PMode pmode = PMode.newBuilder()
+          .setOwnerBits(PMode.parseOwnerBits(mode))
+          .setGroupBits(PMode.parseGroupBits(mode))
+          .setOtherBits(PMode.parseOtherBits(mode))
+          .build();
+      FileOutStream os = mFileSystem.createFile(uri,
+          CreateFilePOptions.newBuilder().setMode(pmode).build());
```
Comment (LuQQiu) [inline integration/fuse/src/main/java/alluxio/fuse/AlluxioFuseFileSystem.java]: PMode is an auto-generated proto file that we are not supposed to touch it. There is an example of how to transfer the input `long mode` to Alluxio `PMode` in https://github.com/Alluxio/alluxio/blob/5f9be23716ef09e2e303931674c66163d613ab58/integration/fuse/src/main/java/alluxio/fuse/AlluxioFuseFileSystem.java#L153

Comment (apc999) [reply integration/fuse/src/main/java/alluxio/fuse/AlluxioFuseFileSystem.java]: re.
In addition, if any static helper methods are needed, it is better to put the helper functions into `alluxio/security/authorization/Mode.java`

File: core/transport/src/main/java/alluxio/grpc/PMode.java (hunk 1 above)
Comment (liuhongtong) [reply core/transport/src/main/java/alluxio/grpc/PMode.java]: Thanks for your comment. I have fixed it.

File: integration/fuse/src/main/java/alluxio/fuse/AlluxioFuseFileSystem.java (hunk 2 above)
Comment (liuhongtong) [reply integration/fuse/src/main/java/alluxio/fuse/AlluxioFuseFileSystem.java]: Good idea. Fixed it.

**Comment by liuhongtong** [comment]
@LuQQiu PTAL again. Thanks.

File: core/transport/src/main/java/alluxio/grpc/PMode.java [hunk 3]
```diff
@@ -674,6 +674,5 @@ public PMode parsePartialFrom(
   public alluxio.grpc.PMode getDefaultInstanceForType() {
     return DEFAULT_INSTANCE;
   }
-
```
Comment (apc999) [inline core/transport/src/main/java/alluxio/grpc/PMode.java]: can you revert the change to this file (the added empty line)?

**Comment by apc999** [comment]
The src code change looks good to me, can you revert the change to the auto-genereted file?
Besides, the unit tests are failing. Can you take a look and see if the tests are already flaky prior to your change or due to the change?

```
Failed tests: 
alluxio.fuse.AlluxioFuseFileSystemTest.flush(alluxio.fuse.AlluxioFuseFileSystemTest)
  Run 1: AlluxioFuseFileSystemTest.flush:189 
Wanted but not invoked:
fileOutStream.flush();
-> at alluxio.fuse.AlluxioFuseFileSystemTest.flush(AlluxioFuseFileSystemTest.java:189)
Actually, there were zero interactions with this mock.
  Run 2: AlluxioFuseFileSystemTest.flush:189 
Wanted but not invoked:
fileOutStream.flush();
-> at alluxio.fuse.AlluxioFuseFileSystemTest.flush(AlluxioFuseFileSystemTest.java:189)
Actually, there were zero interactions with this mock.
alluxio.fuse.AlluxioFuseFileSystemTest.getattrWhenWriting(alluxio.fuse.AlluxioFuseFileSystemTest)
  Run 1: AlluxioFuseFileSystemTest.getattrWhenWriting:283 
Wanted at most 2 times but was 250
  Run 2: AlluxioFuseFileSystemTest.getattrWhenWriting:283 
Wanted at most 2 times but was 249
alluxio.fuse.AlluxioFuseFileSystemTest.write(alluxio.fuse.AlluxioFuseFileSystemTest)
  Run 1: AlluxioFuseFileSystemTest.write:438 
Wanted but not invoked:
fileOutStream.write([42, -128, 1, 3]);
-> at alluxio.fuse.AlluxioFuseFileSystemTest.write(AlluxioFuseFileSystemTest.java:438)
Actually, there were zero interactions with this mock.
  Run 2: AlluxioFuseFileSystemTest.write:438 
Wanted but not invoked:
fileOutStream.write([42, -128, 1, 3]);
-> at alluxio.fuse.AlluxioFuseFileSystemTest.write(AlluxioFuseFileSystemTest.java:438)
Actually, there were zero interactions with this mock.
```

File: core/transport/src/main/java/alluxio/grpc/PMode.java (hunk 3 above)
Comment (liuhongtong) [reply core/transport/src/main/java/alluxio/grpc/PMode.java]: Thanks. reverted.

**Comment by liuhongtong** [comment]
@apc999 Thanks for your comment. All failed test pass.

File: integration/fuse/src/test/java/alluxio/fuse/AlluxioFuseFileSystemTest.java [hunk 4]
```diff
@@ -168,14 +169,17 @@ public void create() throws Exception {
     mFileInfo.flags.set(O_WRONLY.intValue());
     mFuseFs.create("/foo/bar", 0, mFileInfo);
     AlluxioURI expectedPath = BASE_EXPECTED_URI.join("/foo/bar");
-    verify(mFileSystem).createFile(expectedPath);
+    verify(mFileSystem).createFile(expectedPath, CreateFilePOptions.newBuilder()
+        .setMode(new alluxio.security.authorization.Mode((short) 0).toProto())
```
Comment (apc999) [inline integration/fuse/src/test/java/alluxio/fuse/AlluxioFuseFileSystemTest.java]: @LuQQiu  will creating a file with "000" permission bits gets problem? Or, step back, is 0 the right permission level to test?

**Comment by apc999** [comment]
@LuQQiu left one minor comment to get your opinion.

File: integration/fuse/src/test/java/alluxio/fuse/AlluxioFuseFileSystemTest.java (hunk 4 above)
Comment (LuQQiu) [reply integration/fuse/src/test/java/alluxio/fuse/AlluxioFuseFileSystemTest.java]: In local map, I am able to use `unmask 777; touch 1.file` and get a `1.file` without permission.
In mac and Alluxio, 000 are all allowed permission.

Comment (LuQQiu) [reply integration/fuse/src/test/java/alluxio/fuse/AlluxioFuseFileSystemTest.java]: If the given mode is invalid, at least mac will directly reply error message. Invalid mode will not be passed to AlluxioFuse.

**Comment by liuhongtong** [comment]
@apc999 I have verified permission can be set by mkdir. But I have no idea about how mode works yet, since the mode will be ignored.
```java
/**
   * Creates a new dir.
   *
   * @param path the path on the FS of the new dir
   * @param mode Dir creation flags (IGNORED)
   * @return 0 on success, a negative value on error
   */
  @Override
  public int mkdir(String path, @mode_t long mode) {
    final AlluxioURI turi = mPathResolverCache.getUnchecked(path);
    LOG.trace("mkdir({}) [Alluxio: {}]", path, turi);
    try {
      mFileSystem.createDirectory(turi);
    } catch (FileAlreadyExistsException e) {
      LOG.debug("Failed to create directory {}, directory already exists", path);
      return -ErrorCodes.EEXIST();
    } catch (InvalidPathException e) {
      LOG.debug("Failed to create directory {}, path is invalid", path);
      return -ErrorCodes.ENOENT();
    } catch (Throwable t) {
      LOG.error("Failed to create directory {}", path, t);
      return AlluxioFuseUtils.getErrorCode(t);
    }

    return 0;
  }
```

**Comment by apc999** [comment]
can you still patch `mFileSystem.createDirectory(turi);` then for completeness. Then I will merge the PR

**Comment by liuhongtong** [comment]
@apc999 patched mkdir. PTAL

**Comment by apc999** [comment]
alluxio-bot, merge this please

**Comment by apc999** [comment]
alluxio-bot, cherry-pick this to branch-2.0 please

**Comment by apc999** [comment]
alluxio-bot, cherry-pick this to branch-1.8 please

**Comment by liuhongtong** [comment]
@apc999 @LuQQiu Thanks.

**Comment by haoyuan** [comment]
Thanks @liuhongtong for the contribution!
